### PR TITLE
boards: Add support for the Arduino-Leonardo

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Adds support for the ATmega32U4 cpu
- Adds support for the Arduino Leonardo

### Testing procedure

Try to flash and run some tests and examples, such as `examples/hello-world`. Keep in mind that most tests and examples require more resources (RAM & flash) than the Arduino Leonardo has.

### Issues/PRs references

This PR is basically a rebase of https://github.com/RIOT-OS/RIOT/pull/7306

### CODE REVIEW

**Comment by miri64** [comment]
There are some differences between [my branch](https://github.com/miri64/RIOT/tree/leonardo-arduino/feat/initial) (see https://github.com/RIOT-OS/RIOT/pull/7306#issuecomment-461986657) and this PR, so I need to retest. My branch also contains a lot of `BOARD_INSUFFICIENT_MEMORY` sets, so maybe this helps getting this through Murdock ;-).

**Comment by maribu** [comment]
> There are some differences between my branch (see #7306 (comment)) and this PR, so I need to retest.

Sorry for the extra effort. Some refactoring of the AVR based boards has been merged since. Those required adaptations of the code to match current master.

> My branch also contains a lot of BOARD_INSUFFICIENT_MEMORY sets, so maybe this helps getting this through Murdock ;-).

Thanks for the hint. I'll check that out. But if I remember correctly those adaptations have been done prior to the merge of the Arduino Nano - which needed to be added to `BOARD_INSUFFICIENT_MEMORY` for 99% of the apps and tests. I fear that your commit will no longer merge. I'll look into that once this PR is otherwise ready to merge, as a rebase mighy be required to update `BOARD_INSUFFICIENT_MEMORY ` for new tests/apps merged since opening this PR.

**Comment by miri64** [comment]
Sorry for testing only now, but I was quite busy the past month. `examples/hello-world` works, `tests/periph_timer` fails. This was a problem in #7306 as well which was later fixed.

**Comment by maribu** [comment]
@miri64: Thanks for pointing that out. While I tried to deduplicate the timer configs of the ATmega1284P and the ATmega32U4 an error sneaked in (2 instead of 3 channels). I separated the configs now in the fix up, as this seems to be more readable.

I also changed the default frequency in `tests/periph_timer` for the Arduino Leonardo to use the same as for the other ATmega based Arduinos use. With that, the test now passes.

**Comment by miri64** [comment]
I tested both `examples/hello-world` and `tests/periph_timer` again and now both work. Does it make sense to test I²C?

**Comment by maribu** [comment]
On the one hand the I2C driver is the same on all ATmega boards, on the other hand more testing is always better.

Due to memory constraints it will be a bit difficult to test. E.g. the i2c test overflows the memory. [In this branch](https://github.com/maribu/RIOT/tree/i2c_leonardo) I rebased this PR against master (to get the i2c_scan shell command) and added the `examples/i2c_scan` as a minimal RIOT shell with only the commands `reboot` and `i2c_scan` provided. This is small enough to fit the Arduino Leonardo and allows very basic testing of the I2C interface.

**Comment by miri64** [comment]
Tested I²C with `tests/drivers_ds1307` (by attaching a ds1307 to the I²C pins first of course). Works :+1:.

**Comment by maribu** [comment]
Perfect :-) Thanks :-)

**Comment by miri64** [comment]
Some builds already fail due to memory contstaints

**Comment by maribu** [comment]
Yes, all the `BOARD_INSUFFICIENT_MEMORY` need to be extended. Will take about 30 minutes to do...

**Comment by miri64** [comment]
> Yes, all the `BOARD_INSUFFICIENT_MEMORY` need to be extended. Will take about 30 minutes to do...

sure. I remember the pain from #7306 ;-)

**Comment by maribu** [comment]
> his branch has conflicts that must be resolved

F**k. (These are the `BOARD_INSUFFICIENT_MEMORY` updates). May I rebase?

**Comment by miri64** [comment]
Of course!

**Comment by miri64** [comment]
`arduino-leonardo` also needs to be blacklisted for 3 application (two because no `malloc.h` was found, one for because *our* `sys/types.h` for avrlibc defines `time_t` wrong [will fix]).

**Comment by miri64** [comment]
Ok... apparently making `time_t` signed leads to problems `avrlibc` (see https://github.com/RIOT-OS/RIOT/pull/8857). So won't fix.

**Comment by miri64** [comment]
But for reference, in [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799.2018edition/basedefs/sys_types.h.html#tag_13_67) it shall be an integer type.

**Comment by miri64** [comment]
@maribu you may squash. The blacklisting of the tests should be done in a separate commit anyways.

**Comment by miri64** [comment]
You may squash the missing blacklisting for `tests/pipe` immediately. Let's finally get this in!

**Comment by miri64** [comment]
:tada:

**Comment by maribu** [comment]
@tprrt: Thank you very much for implementing the Arduino Leonardo support!

@miri64: Thanks for the review and for noticing my mess up with the timers :-)

**Comment by maribu** [comment]
@miri64:

> But for reference, in [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799.2018edition/basedefs/sys_types.h.html#tag_13_67) it shall be an integer type.

I think that integer type does imply whether it is signed or unsigned. E.g. for `short int`, `long int`, `long long int` it is clear that they are signed, but `char` is also an integer type, but can either by signed or unsigned. (You can use `signed char` and `unsigned char` if you need to know what you get.)

**Comment by miri64** [comment]
> I think that integer type does imply whether it is signed or unsigned. E.g. for `short int`, `long int`, `long long int` it is clear that they are signed, but `char` is also an integer type, but can either by signed or unsigned. (You can use `signed char` and `unsigned char` if you need to know what you get.)

You are right. The value `mktime()` (and in turn `timegm()`) shall return is `(time_t)-1`, so it is actually `minmea` that needs a fix.

**Comment by miri64** [comment]
Forgot [the reference](http://pubs.opengroup.org/onlinepubs/9699919799/functions/mktime.html).

**Comment by miri64** [comment]
Apparently this [is fixed in the current version](https://github.com/kosma/minmea/blob/ae4dd9442a9041345d5ef108f062e7e4ec6954f2/minmea.c#L635).

File: examples/default/Makefile [hunk 1]
```diff
@@ -31,7 +32,6 @@ QUIET ?= 1
 
 # Modules to include:
 USEMODULE += shell
-USEMODULE += shell_commands
```
Comment (miri64) [inline examples/default/Makefile]: Why was this removed? oO

Comment (miri64) [reply examples/default/Makefile]: Fixed in https://github.com/RIOT-OS/RIOT/pull/11599

Comment (maribu) [reply examples/default/Makefile]: I have no idea how that sneaked in. I did touch about every example and test Makefile to update the `BOARD_INSUFFICIENT_MEMORY`, so I guess I just hit a wrong key and lost track of the changes :-(

Comment (miri64) [reply examples/default/Makefile]: Me as well... Shit happens ;-).
